fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java

```diff
@@ -135,56 +162,76 @@ public static ScalarOperator translate(Expr expression, ExpressionMapping expres
 
     public static ScalarOperator translate(Expr expression) {
         IgnoreSlotVisitor visitor = new IgnoreSlotVisitor();
-        return visitor.visit(expression, null);
+        return visitor.visit(expression, new Context());
     }
 
-    public static ScalarOperator translateWithoutRewrite(Expr expression, ExpressionMapping expressionMapping,
-                                                         ColumnRefFactory columnRefFactory) {
-        ColumnRefOperator columnRefOperator = expressionMapping.get(expression);
-        if (columnRefOperator != null) {
-            return columnRefOperator;
-        }
+    private static final class Context {
 
-        Visitor visitor = new Visitor(expressionMapping, new ArrayList<>(), columnRefFactory);
-        ScalarOperator result = visitor.visit(expression, null);
+        public final boolean useSemiAnti;
+        public final List<Expr> outerExprs;
 
-        ScalarOperatorRewriter scalarRewriter = new ScalarOperatorRewriter();
-        result = scalarRewriter.rewrite(result, ScalarOperatorRewriter.DEFAULT_TYPE_CAST_RULE);
+        public Context() {
+            useSemiAnti = false;
+            outerExprs = Collections.emptyList();
+        }
 
-        requireNonNull(result, "translated expression is null");
-        return result;
+        public Context(boolean useSemiAnti, List<Expr> outerExprs) {
```

Comment:
Why all expr need to consider `useSemiAnti` or `outerExprs`, which seems not reasonable.

Reply:
The context is only used for subquery

Reply:
But all exprs use this class.